File: fb/121.best-time-to-buy-and-sell-stock.cpp
// Time O(n)
// Space O(1)
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if(prices.empty()) return 0;
        int ans = 0;
        for(int i = 1, min = prices[0]; i < prices.size(); i++) {
            if(prices[i] < min) {
                min = prices[i];
            }
            else {
                ans = max(ans, prices[i] - min);
            }
        }
        return ans;
    }
};

File: fb/121.best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if prices is None:
            return 0
        
        maxProfit = 0
        minPrice = prices[0]
        for i in range(1, len(prices)):
            if minPrice > prices[i]:
                minPrice = prices[i]
            else:
                maxProfit = max(prices[i] - minPrice, maxProfit)
                
        return maxProfit
